Bring feature order and rendering order together in a useful way

joto opened the pull request:
There are two ways of taking "order" into account when rendering:

1. You can control the feature rendering order which essentially does an "order by" of your layer data.
2. You can enable symbol levels which defines which styles are rendered in which order.

This commit addresses a problem when trying to bring (1) and (2) together. When both are enabled, the rendering order (2) trumps the order given by (1), essentially ignoring the order from (1).

With this commit rendering instead starts "from scratch" every time the "order by" attributes have a new value.

This allows, for instance, to render complex roads networks correctly. Ordering via (1) is used to draw roads going over bridges etc. in the correct order while ordering via (2) is used to render road styles correctly (thicker dark line, then thinner light line to get a roads with casing).

The patch works by putting all features that need to be rendered not in a single QHash (Symbol->Feature) but a QList of those hashes. Every time at least one of the attributes used for the "order by" changes, a new element is added to that list. The patch is rather simple, it just looks a bit larger due to the extra indentation needed.

I thought about making this configurable, but I don't see any use case where it makes sense to have (1) and (2) enabled where you don't want this behaviour.

Fixes #42428
See #17276

<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - You have run the `scripts/prepare_commit.sh` script (https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit.
    If you didn't do this, you can also run `./scripts/astyle_all.sh` from your source folder.

  - You have read the QGIS Coding Standards (https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
-->

joto committed:
Bring feature order and rendering order together in a useful way

There are two ways of taking "order" into account when rendering:

1. You can control the feature rendering order which essentially does an
   "order by" of your layer data.
2. You can enable symbol levels which defines which styles are rendered
   in which order.

This commit addresses a problem when trying to bring (1) and (2)
together. When both are enabled, the rendering order (2) trumps the
order given by (1), essentially ignoring the order from (1).

With this commit rendering instead starts "from scratch" every time the
"order by" attributes have a new value.

This allows, for instance, to render complex roads networks correctly.
Ordering via (1) is used to draw roads going over bridges etc. in the
correct order while ordering via (2) is used to render road styles
correctly (thicker dark line, then thinner light line to get a roads
with casing).

The patch works by putting all features that need to be rendered not in
a single QHash (Symbol->Feature) but a QList of those hashes. Every time
at least one of the attributes used for the "order by" changes, a new
element is added to that list. The patch is rather simple, it just looks
a bit larger due to the extra indentation needed.

I thought about making this configurable, but I don't see any use case
where it makes sense to have (1) and (2) enabled where you don't want
this behaviour.

Fixes #42428
See #17276

joto commented:
![road-layering](https://user-images.githubusercontent.com/113756/222926489-533296df-e503-47ea-acd8-414b7938eb5d.png)

Here you can see the change in action. The layering is all correct and only one QGIS layer was used.

m-kuhn commented:
Thank you for opening this pull request.
The idea / logic sounds good to me and this use case will help produce nice cartographic results.

Could you do some syntactic changes to the code:

 - use `camelCase` variables
 - add a comment that explains the semantics of the new `QList` that was added (i.e. replace the term "array" with an explanation of what this list is used for)
 - add a comment that explains the meaning of the empty items in the list

These comments  will help the review process as well as for future maintenance.

It would also be nice to have a simple rendering test that covers this scenario. (see e.g. here: https://github.com/qgis/QGIS/blob/master/tests/src/core/testqgs25drenderer.cpp).

m-kuhn commented on src/core/vector/qgsvectorlayerrenderer.cpp:
```diff
@@ -542,7 +542,28 @@ void QgsVectorLayerRenderer::drawRendererLevels( QgsFeatureRenderer *renderer, Q
   const bool isMainRenderer = renderer == mRenderer;
 
-  QHash< QgsSymbol *, QList<QgsFeature> > features; // key = symbol, value = array of features
+  // We need to figure out in which order all the features should be rendered.
+  // Ordering is based on (a) a "level" which is determined by the configured
+  // feature rendering order" and (b) the symbol level. The "level" is
+  // determined by the values of the attributes defined in the feature
+  // rendering order settings. Each time the attribute(s) have a new distinct
+  // value, a new empty QHash is added to the "features" list. This QHash is
+  // then filled by mappings from the symbol to a list of all the features
+  // that should be rendered by that symbol.
+  //
+  // If orderBy is not enabled, this list will only ever contain a single
+  // element.
```
thanks :heart:

joto commented:
Thanks for your feedback. I have fixed the variable names and added some comments to the code. I need some more time to look at the tests and see what I need to do.

One potential problem I noticed while looking at the code again: If the attribute we do the *order by* on has many values, maybe even a different value for each feature, we'll generate an aweful lot of `QList` entries with one-element `QHash`es in them. This can use up a lot of memory. I don't think there is a way around that, and a user shouldn't do that, but if they do, performance will suffer. So maybe we have to detect that case in some way and handle it specially.

joto committed:
Add tests for using ordered layers and symbol levels together

joto commented:
@m-kuhn Any chance to get this reviewed?

m-kuhn commented:
> One potential problem I noticed while looking at the code again: If the attribute we do the _order by_ on has many values, maybe even a different value for each feature, we'll generate an aweful lot of `QList` entries with one-element `QHash`es in them. This can use up a lot of memory. I don't think there is a way around that, and a user shouldn't do that, but if they do, performance will suffer. So maybe we have to detect that case in some way and handle it specially.

This is a fair point. Could you conduct a test with 10000 features to check the effect?

Another thing is the usage of std containers, compared to Qt containers which are used in the majority of cases. I am not sure if there's a guideline for that, so I'll leave it to others to speak up if there's an issue with that.

Apart from this, it looks good to me.

joto committed:
Use QVector instead of std::vector

joto commented:
> This is a fair point. Could you conduct a test with 10000 features to check the effect?

I tested it with a layer of >500,000 features using the unique id of the feature as "layer" (so it starts a new "layer" for each feature making this the worst case). It used more memory, but not that much and rendering speed wasn't much worse either, so I think we are good.

> Another thing is the usage of std containers, compared to Qt containers which are used in the majority of cases. I am not sure if there's a guideline for that, so I'll leave it to others to speak up if there's an issue with that.

I switched from `std::vector` to `QVector`.

m-kuhn reviewed (approved):
Thank you

JohnProv commented:
Does this patch also concern rule-based rendering?
In 3.32.0 it works correctly with the categorized renderer, but not with the rule-based renderer. There the problem from #42428 still exists.

joto commented:
@JohnProv I just checked this and you are right, it works in the categorized renderer, but not with the rule-based renderer. As far as I can see there is nothing specific to the renderer type in the code I changed. But I know next to nothing about the QGIS internals, just beat at the code until it worked for me. Clearly I missed something there. I suggest you reopen #42428 and put a comment there that it isn't fixed after all.
